LibreNMS/IPMI/IPMIClient.php

```diff
@@ -0,0 +1,168 @@
+    {
+        $this->ipmiToolPath = $ipmiToolPath;
+        $this->host = $host;
+        $this->user = $user;
+        $this->password = $password;
+    }
+
+    public function getHost(): string
+    {
+        return $this->host;
+    }
+
+    /**
+     * Gets the IPMI interface driver used by the client.
+     */
+    public function getDriver(): string
+    {
+        return $this->interface;
+    }
+
+    /**
+     * Sets the IPMI interface driver.
+     */
+    public function setDriver(string $interface)
+    {
+        $this->interface = $interface;
+    }
+
+    /**
+     * Gets the port used by the client.
+     * @return ?string The port currently used. Null if not specified.
+     */
+    public function getPort(): ?string
+    {
+        return $this->port;
+    }
+
+    /**
+     * Set the port used by the client.
+     */
+    public function setPort(?string $port)
+    {
+        $this->port = $port;
+    }
+
+    /**
+     * Gets a binary representation of the cached SDR record for this host.
+     * @return string|false The SDR binary or false on failure.
+     */
+    public function getSDR()
+    {
+        $path = '/tmp/librenms/ipmi/SDR';
+        $this->assertSDR($path);
+
+        $filename = "$path/$this->host";
+        $sdrFile = fopen($filename, 'r');
+        try {
+            return fread($sdrFile, filesize($filename));
+        } finally {
+            fclose($sdrFile);
+        }
+    }
+
+    /**
+     * Gets a list of sensors and threshold values reported by ipmitool.
+     */
+    public function getSensors(): array
+    {
+        return explode(PHP_EOL, $this->send('sensor'));
+    }
+
+    /**
+     * Gets a comma-separated list of sensor values from the
+     * Sensor Data Repository (SDR).
+     */
+    public function getSensorDataRepository(): array
+    {
+        return explode(PHP_EOL, $this->send('-c sdr'));
+    }
+
+    /**
+     * Sends an ipmitool command with specified parameters.
+     * @param string $command the command to send.
+     * @param bool $escalatePrivileges a boolean indicating whether to use 'USER' or 'ADMINISTRATOR' privilege.
+     * @return null|string The stdout of the command as reported by ipmitool.
+     */
+    public function sendCommand(string $command, bool $escalatePrivileges = false): ?string
+    {
+        return $this->send($command, $escalatePrivileges);
+    }
+
+    private function send(string $command, bool $escalate = false)
+    {
+        $cmd = [$this->ipmiToolPath];
+        if ($this->host != 'localhost') {
+            array_push($cmd, '-H', $this->host, '-U', $this->user, '-P', $this->password);
+            array_push($cmd, '-L', $escalate ? 'ADMINISTRATOR' : $this->privLvl);
+            if ($this->port) {
+                array_push($cmd, '-p', $this->port);
+            }
+
+            array_push($cmd, '-I', $this->interface);
+        }
+
+        $cmd = array_merge($cmd, explode(' ', $command));
+
+        return external_exec($cmd);
```

Comment:
Are you sure you want to use this function? It's very snmp centric; https://github.com/librenms/librenms/blob/5ba3df45eb5a351cec7f52a05c26d4936dcf4d0f/includes/common.php#L69

Reply:
Good point. I will create a slimmer version for ipmi execution. That way, I can also check and abort if the command errored.

Reply:
You can use https://github.com/librenms/librenms/blob/master/LibreNMS/Proc.php

Reply:
Resolved in 03a28a6f81537fa7a1f44587bc20bb5c4baab9da.